source/common/quic/active_quic_listener.cc

```diff
@@ -44,8 +44,10 @@ ActiveQuicListener::ActiveQuicListener(
       kernel_worker_routing_(kernel_worker_routing),
       packets_to_read_to_connection_count_ratio_(packets_to_read_to_connection_count_ratio),
       crypto_server_stream_factory_(crypto_server_stream_factory),
-      connection_id_generator_(std::move(cid_generator)) {
+      connection_id_generator_(std::move(cid_generator)),
+      select_connection_id_worker_(std::move(worker_selector)) {
   ASSERT(!GetQuicFlag(quic_header_size_limit_includes_overhead));
+  ASSERT(select_connection_id_worker_ != nullptr);
```

Comment:
I would think it be okay to pass an empty function here if the extension uses kernel routing.

Reply:
Not if this function is the routing fallback for packets redirected during hot restart.

Reply:
What gets passed in to this constructor is the BPF compatible impl of WorkerSelector. And what the listener needs during hot restart is one that knows about the forked instance. I feel that it is okay not to have a worker selector if the Envoy is configured to use kernel routing until it gets into hot restart state. So this constructor can take nullptr, and the hot restart callback should pass in a handle to a worker selector. WDYT?

Reply:
That sounds pretty painful - that would require the hot restart code to figure out the appropriate ConnectionIdGeneratorFactory to use, the concurrency value, etc. and then updating all the listeners, *and* still having this same branch anyway for the case when bpf isn't available. Versus this way everything that's needed is all right here when the bpf program is (or isn't) being initialized.

I don't think the listener needs to know anything about the forked instance or hot restart for routing received packets - the worker id should resolve the same way for the same packet. It only needs to know about hot restart for forwarding undeliverable packets (which isn't this PR).

Reply:
I think it's simpler to make it always have a valid selector function; if we're using ebpf it may never be called, but the code is simpler if it's required anyways.

Reply:
Can you give some pseudo code about how to extend a worker selector which does BPF compatible routing for most of the time and packet forwarding during hot restart? I feel some part of the implementation will be shared with all extensions.

Reply:
You don't extend the worker selector, the worker selector is just for finding the right worker id, which is, say, 12 for both the parent and child process. The part that's relevant to hot restart is that when worker 12 on the parent fails to deliver the packet, it gets intercepted by the hot restart packet handler, that handler sends the packet to the child process, and the child process can use the worker selector to figure out that the packet should go to worker 12; it can't use bpf for that.

We *could* just encapsulate which worker it should be delivered to in the packet forwarding RPC, in which case the worker selector function is only for if bpf isn't available. Even then it would be significantly more complicated (especially for testing) to have the function only populated when bpf isn't available, rather than populated and unused when bpf is available.

(One downside of that, also, would be that it would be incompatible with a change in concurrency between instances, but as I understand it we don't support hot restart under those conditions anyway.)

Reply:
Okay, my understanding was wrong. I thought the worker selector will forward packets in parent instance during hot restart. Instead, it is used in child instance to route packets received via RPC. 

Enforcing this callback not null makes sense.